platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/tipset/TipsetScoreCalculator.java

```diff
@@ -0,0 +1,281 @@
+        if (eventTipset == null) {
+            throw new IllegalArgumentException("event " + event + " is not in the tipset tracker");
+        }
+
+        final TipsetAdvancementWeight score = snapshot.getTipAdvancementWeight(selfId, eventTipset);
+        if (score.advancementWeight() > maximumPossibleAdvancementWeight) {
+            throw new IllegalStateException("score " + score + " is greater than the maximum possible score "
+                    + maximumPossibleAdvancementWeight);
+        }
+
+        final TipsetAdvancementWeight scoreImprovement = score.minus(previousScore);
+
+        if (isSuperMajority(score.advancementWeight() + selfWeight, totalWeight)) {
+            snapshot = eventTipset;
+            snapshotHistory.add(snapshot);
+            if (snapshotHistory.size() > maxSnapshotHistorySize) {
+                snapshotHistory.remove();
+            }
+            previousScore = ZERO_ADVANCEMENT_WEIGHT;
+        } else {
+            previousScore = score;
+        }
+
+        latestSelfEventTipset = eventTipset;
+
+        return scoreImprovement;
+    }
+
+    /**
+     * Figure out what advancement score we would get if we created an event with a given list of parents.
+     *
+     * @param parents the proposed parents of an event
+     * @return the advancement score we would get by creating an event with the given parents
+     */
+    public TipsetAdvancementWeight getTheoreticalAdvancementScore(@NonNull final List<EventDescriptor> parents) {
+        if (parents.isEmpty()) {
+            return ZERO_ADVANCEMENT_WEIGHT;
+        }
+
+        final List<Tipset> parentTipsets = new ArrayList<>(parents.size());
+        for (final EventDescriptor parent : parents) {
+            parentTipsets.add(tipsetTracker.getTipset(parent));
+        }
+
+        // Don't bother advancing the self generation in this theoretical tipset,
+        // since self advancement doesn't contribute to tipset score.
+        final Tipset newTipset = Tipset.merge(parentTipsets);
+
+        return snapshot.getTipAdvancementWeight(selfId, newTipset).minus(previousScore);
+    }
+
+    /**
+     * Compute the current maximum bully score with respect to all nodes. This is a measure of how well slow nodes'
+     * events are being incorporated in the hashgraph by faster nodes. A high score means slow nodes are being bullied
+     * by fast nodes. A low score means slow nodes are being included in consensus. Lower scores are better.
+     *
+     * @return the current tipset bully score
+     */
+    public int getMaxBullyScore() {
+        int bullyScore = 0;
+        for (final EventDescriptor eventDescriptor : childlessEventTracker.getChildlessEvents()) {
+            bullyScore = Math.max(bullyScore, getBullyScoreForNode(eventDescriptor.getCreator()));
+        }
+        return bullyScore;
+    }
+
+    /**
+     * Get the bully score with respect to one node. A high bully score means that we have access to events that could
```

Comment:
I think adding this will help even further:

```suggestion
     * Get the bully score with respect to one node, i.e. how much this node is bullying the specified node. A high bully score means that we have access to events that could
```

Reply:
updated